### PR TITLE
Feature: ResourceBuilder Update

### PR DESCRIPTION
**Pull request by talekhinezh**
## Summary
Refactor Scrypto ResourceBuilder to use the more consistent role setting interface

## Details
* Added `mintable!`, `burnable!`, `restrict_withdraw!`, `restrict_deposit!`, `freezable!`, `recallable!` macros to be used with ResourceBuilder
* Refactored ResourceManager create functions to be more explicit about resource manager configuration

### For dApp Developers
The same role setting interface is now used with the resource builder:
```
ResourceBuilder::new_integer_non_fungible::<Sandwich>(OwnerRole::Fixed(allow_all))
  .mint_roles(mint_roles! {
    minter => OWNER, locked;
    minter_updater => rule!(deny_all), locked;
  })
  .burn_roles(burn_roles! {
    burner => rule!(require(XRD)), updatable;
    burner_updater => OWNER, locked;
  })
  .create_with_no_initial_supply();
```

### For Internal Integrators
The ResourceManager create function interfaces have been updated to be more explicit about configuration

### CODE REVIEW

**Review by iamyulong — COMMENTED** [review]
Can we rename `resource_action`/`ResourceAction` to `resource_feature`/`ResourceFeature`?

Also, note that we're not checking if the role definitions match the enabled resource feature.

From this, I feel that I like the old model a bit more, because it's more structured (another abstraction in resource system; don't have to be aligned with globalize system API).

File: radix-engine/src/blueprints/resource/resource_manager_common.rs [hunk 1]
```diff
@@ -225,15 +194,26 @@ pub fn globalize_non_fungible_with_initial_supply<Y>(
     owner_role: OwnerRole,
     object_id: NodeId,
     resource_address_reservation: GlobalAddressReservation,
-    access_rules: BTreeMap<ResourceAction, (AccessRule, AccessRule)>,
+    mut main_roles: RolesInit,
```
Comment (iamyulong) [inline radix-engine/src/blueprints/resource/resource_manager_common.rs:197]: The way we converts `NonFungibleResourceFeatures` into features and roles, and then `verify_and_defaults_to_main_roles`, seems unnecessary?

One of the benefits of using `NonFungibleResourceFeatures` to have structured data and reduce validation needed.

File: radix-engine-interface/src/blueprints/resource/fungible/fungible_resource_manager.rs [hunk 2]
```diff
@@ -4,21 +4,131 @@ use crate::*;
...
+                recaller_updater => rule!(deny_all), locked;
+            },
+            restrict_withdraw: restrict_withdraw! {
+                withdrawer => access_rule.clone(), locked;
+                withdrawer_updater => rule!(deny_all), locked;
+            },
+            restrict_deposit: restrict_deposit! {
+                depositor => access_rule.clone(), locked;
+                depositor_updater => rule!(deny_all), locked;
+            },
```
Comment (iamyulong) [inline radix-engine-interface/src/blueprints/resource/fungible/fungible_resource_manager.rs:52]: Is restricting deposit and withdraw intended?

File: radix-engine-interface/src/blueprints/resource/fungible/fungible_resource_manager.rs [hunk 3]
```diff
@@ -4,21 +4,131 @@ use crate::*;
...
+        let mut features = Vec::new();
+        let mut roles = RolesInit::new();
+
+        if let Some(mintable) = &self.mintable {
+            if mintable
+                .minter
+                .ne(&RoleDefinition::locked(AccessRule::DenyAll))
+            {
+                features.push(MINT_FEATURE);
+            }
```
Comment (iamyulong) [inline radix-engine-interface/src/blueprints/resource/fungible/fungible_resource_manager.rs:66]: Premature optimisation?

Comment (talekhinezh) [reply radix-engine-interface/src/blueprints/resource/fungible/fungible_resource_manager.rs:66]: Yup, this has been fixed in another branch but will just merge that branch into this one

File: assets/blueprints/genesis_helper/src/lib.rs [hunk 4]
```diff
@@ -263,10 +263,16 @@ mod genesis_helper {
                 })
                 .with_address(resource.address_reservation);
 
-            if let Some(address) = owner_badge_address {
+            if owner_badge_address.is_some() {
                 builder
-                    .mintable(rule!(require(address)), rule!(deny_all))
-                    .burnable(rule!(require(address)), rule!(deny_all))
+                    .mintable(mintable! {
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply assets/blueprints/genesis_helper/src/lib.rs:271]: we do, it's specified in the owner role

File: radix-engine/src/blueprints/resource/resource_manager_common.rs [hunk 5]
```diff
@@ -1,145 +1,93 @@
...
-            main_roles.define_role(DEPOSITOR_ROLE, deposit_access_rule, locked);
-            main_roles.define_role(DEPOSITOR_UPDATER_ROLE, deposit_mutability, locked);
-        }
-    }
-
-    main_roles
+    Ok(())
 }
 
 pub fn features(
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/blueprints/resource/resource_manager_common.rs:67]: good catch

File: radix-engine/src/blueprints/resource/resource_manager_common.rs (hunk 1 above)
Comment (talekhinezh) [reply radix-engine/src/blueprints/resource/resource_manager_common.rs:197]: You're right, the old `verify_and_defaults_to_main_roles` should have been removed.

File: radix-engine/src/blueprints/resource/resource_manager_common.rs [hunk 6]
```diff
@@ -183,15 +141,26 @@ pub fn globalize_fungible_with_initial_supply<Y>(
...
+        if object_id.is_global_fungible_resource_manager() {
+            RuntimeError::ApplicationError(ApplicationError::FungibleResourceManagerError(
+                FungibleResourceManagerError::InvalidRole(err),
+            ))
+        } else {
+            RuntimeError::ApplicationError(ApplicationError::NonFungibleResourceManagerError(
+                NonFungibleResourceManagerError::InvalidRole(err),
+            ))
+        }
+    })?;
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/blueprints/resource/resource_manager_common.rs:92]: This is true, can we save this cleanup for the next PR?

[the remark this answers is not in the document]
Comment (iamyulong) [reply radix-engine/src/blueprints/resource/resource_manager_common.rs:92]: Sure!

**Review by iamyulong — APPROVED** [review]

**Pull request merged**